lvol - bug fix - Convert units to lowercase when using LVS or VGS command

zigaSRC opened the pull request:
##### SUMMARY
Converted units to lowercase when using LVS or VGS command. 

Fixes #2360
Fixes #2370

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
lvol

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```paste below

```

felixfontein commented:
What about all the places where `size_unit` itself is used?

zigaSRC commented:
Doesn't matter since the other tools don't differentiate lower and upper case units. Do you think it's best to use the lower case everywhere? 

And since we're already talking it might also be best to use the default unit everywhere when a different unit is not specified. The default unit `m` is only applied to the `unit` variable on line 385, but from that point on there's plenty of references to the size_unit variable which is empty in that case.

quidame commented on plugins/modules/system/lvol.py:
```diff
@@ -389,7 +389,7 @@ def main():
     vgs_cmd = module.get_bin_path("vgs", required=True)
     rc, current_vgs, err = module.run_command(
-        "%s --noheadings --nosuffix -o vg_name,size,free,vg_extent_size --units %s --separator ';' %s" % (vgs_cmd, unit, vg))
+        "%s --noheadings --nosuffix -o vg_name,size,free,vg_extent_size --units %s --separator ';' %s" % (vgs_cmd, unit.lower(), vg))
```
Why not to enhance this same piece of code until the end ?
```suggestion
        [vgs_cmd, "--noheadings", "--nosuffix",  "-o", "vg_name,size,free,vg_extent_size", "--units", unit.lower(), "--separator", ";", vg])

zigaSRC replied:
What do you mean by this? As far as I see you only changed the syntax to use a list. 

I'm not familiar with coding ansible modules; Is this the recommended way of calling module.run_command?

quidame replied:
yes, it is: https://docs.ansible.com/ansible/latest/reference_appendices/module_utils.html
> args – is the command to run * If args is a list, the command will be run with shell=False. * If args is a string and use_unsafe_shell=False it will split args to a list and run with shell=False * If args is a string and use_unsafe_shell=True it runs with shell=True.

felixfontein replied:
It is, but let's not do too much in this bugfix and instead get it out faster. Such improvements can wait for afterwards.

quidame commented on plugins/modules/system/lvol.py:
```diff
@@ -403,7 +403,7 @@ def main():
     lvs_cmd = module.get_bin_path("lvs", required=True)
     rc, current_lvs, err = module.run_command(
-        "%s -a --noheadings --nosuffix -o lv_name,size,lv_attr --units %s --separator ';' %s" % (lvs_cmd, unit, vg))
+        "%s -a --noheadings --nosuffix -o lv_name,size,lv_attr --units %s --separator ';' %s" % (lvs_cmd, unit.lower(), vg))
```
also here
```suggestion
        ["lvs_cmd", "-a", "--noheadings", "--nosuffix", "-o", "lv_name,size,lv_attr", "--units", unit.lower(), "--separator", ";", vg])
```

zigaSRC commented:
Added fix for #2370.

After some more research into the lvm documentation, it appears there were a few changes over the versions. The latest docs I could find set the given value (with -l option) as the maximum meaning they always round down the value. I changed the implementation to match this behavior. 

In my tests subtracting (for example 10%VG) and then expanding the logical volume size with the `+` and `-` options will fail with one extent of space missing. This is intended as doing the same with the tools themselves gives the same result. 

I apologize for all the back and forth and the introduction of bugs on my part.

felixfontein commented:
@MalfuncEddie @dominikkaminski @jake2184 @quidame @aminvakil I'm planning on releasing a bugfix release 3.0.1 tomorrow, and I need to decide whether this can get merged for it or not. Can you review, and/or test whether it fixes the issues you are experiencing?

MalfuncEddie commented:
> @MalfuncEddie @dominikkaminski @jake2184 @quidame @aminvakil I'm planning on releasing a bugfix release 3.0.1 tomorrow, and I need to decide whether this can get merged for it or not. Can you review, and/or test whether it fixes the issues you are experiencing?

Test it today (was sick last couple of days)

dominikkaminski commented:
> @MalfuncEddie @dominikkaminski @jake2184 @quidame @aminvakil I'm planning on releasing a bugfix release 3.0.1 tomorrow, and I need to decide whether this can get merged for it or not. Can you review, and/or test whether it fixes the issues you are experiencing?

I tested #2370 and this PR fixes the issue. All tasks run now as intended again.

Thanks `@all` for the fast Bugfix and all the investigation around it.

MalfuncEddie commented:
Hi,

Juist tested this and I still have issues
```yml
cat test0.yml
- name: test
  become: true
  become_method: sudo
  gather_facts: no
  hosts: all
  tasks:
    - name: 10G /tmp
      community.general.lvol:
        vg: rootvg
        lv: tmp
        size: "10G"
        resizefs: yes

    - name: 10G /tmp
      community.general.lvol:
        vg: rootvg
        lv: tmp
        size: "10G"
        resizefs: yes
```
`ansible-playbook -i ,172.31.222.227  -k -K test0.yml
SSH password:
BECOME password[defaults to SSH password]:

PLAY [test] ****************************************************************************************************************************************************************************************

TASK [10G /tmp] ************************************************************************************************************************************************************************************
changed: [172.31.222.227]

TASK [10G /tmp] ************************************************************************************************************************************************************************************
fatal: [172.31.222.227]: FAILED! => {"changed": false, "msg": "Sorry, no shrinking of tmp without force=yes."}
        to retry, use: --limit @/home/users/axiroot/test0.retry

PLAY RECAP *****************************************************************************************************************************************************************************************
172.31.222.227             : ok=1    changed=1    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0
`

zigaSRC commented:
Could you provide a bit more info about your system and the logical volume? 
I could not reproduce the error on my system with the latest version of the lvol.py from this PR.

MalfuncEddie commented:
target system is an oracle enterprise linux 8.3 (latest patches) (basic os install)
`
vgs
  VG     #PV #LV #SN Attr   VSize   VFree
  rootvg   1   8   0 wz--n- <43.00g <13.50g
lvs
  LV            VG     Attr       LSize   Pool Origin Data%  Meta%  Move Log Cpy%Sync Convert
  home          rootvg -wi-ao---- 512.00m
  root          rootvg -wi-ao----  10.00g
  swap          rootvg -wi-ao----   4.00g
  tmp           rootvg -wi-ao----  10.00g
  var           rootvg -wi-ao----   2.00g
  var_log_audit rootvg -wi-ao---- 512.00m
  var_tmp       rootvg -wi-ao---- 512.00m


`

MalfuncEddie commented:
@zigaSRC 
I thnik sonething is wierd on my ansible controller and was using the old version (3.0.0) instead of the bugfix. Migrating to an an other controller and and running this branch gave no problem.

So okay for me!

Sorry for the confusion

felixfontein commented:
@MalfuncEddie heh, that sometimes happens :) I'm glad the PR solves the issue for you as well. I guess we can merge it then. If nobody complains, I'll do that this evening / tomorrow morning.

felixfontein commented:
@zigaSRC thanks for fixing this!
@aminvakil @quidame @MalfuncEddie @dominikkaminski thanks for reviewing and testing!
